drivers/python/rethinkdb/net_twisted.py

```diff
@@ -0,0 +1,417 @@
+        p = DatabaseProtocol(self)
+        return p
+
+    def clientConnectionLost(self, connector, reason):
+        pass
+
+    def clientConnectionFailed(self, connector, reason):
+        pass
+
+class CursorItems(DeferredQueue):
+
+    def __init__(self):
+        super(CursorItems, self).__init__(self)
+
+    @inlineCallbacks
+    def wait_for_new_item(self):
+        """
+        Callback whenever a new item arrived.
+        It use the behavior of put.
+        Warning: It ignores QueueUnderflow / QueueOverflow mechanisms.
+        """
+        d = Deferred(canceller=self._cancelGet)
+        self.waiting.append(d)
+        item = yield d
+        self.pending.append(item)
+        returnValue(item)
+
+    def cancel_getters(self, err):
+        """
+        Cancel all waiters.
+        """
+        for waiter in self.waiting[:]:
+            waiter.errback(err)
+            self.waiting.remove(waiter)
+
+    def __len__(self):
+        return len(self.pending)
+
+    def __getitem__(self, key):
+        return self.pending[key]
+
+class TwistedCursor(Cursor):
+
+    def __init__(self, *args, **kwargs):
+        super(TwistedCursor, self).__init__(*args, **kwargs)
+        self.items = CursorItems()
+        self.new_response = Deferred()
+
+    def _extend_queue(self, data):
+        for k in data:
+            self.items.put(k)
+
+    def _extend(self, res):
+        self.outstanding_requests -= 1
+        self.threshold = len(res.data)
+        if self.error is None:
+            if res.type == pResponse.SUCCESS_PARTIAL:
+                self._extend_queue(res.data)
+            elif res.type == pResponse.SUCCESS_SEQUENCE:
+                self._extend_queue(res.data)
+                self.error = self._empty_error()
+                self.items.cancel_getters(self.error)
+            else:
+                self.error = res.make_error(self.query)
+            self._maybe_fetch_batch()
+
+        if self.outstanding_requests == 0 and self.error is not None:
+            del self.conn._cursor_cache[res.token]
+
+        self.new_response.callback(None)
+        self.new_response = Deferred()
+
+    def _empty_error(self):
+        return RqlCursorEmpty(self.query.term)
+
+    @inlineCallbacks
+    def fetch_next(self, wait=True):
+        timeout = Cursor._wait_to_timeout(wait)
+        start_timeout(timeout, self.new_response)
+        while len(self.items) == 0 and self.error is None:
+            self._maybe_fetch_batch()
+            try:
+                yield self.new_response
+            except CancelledError:
+                raise RqlTimeoutError()
+
+        returnValue(not self.is_empty() or self.has_error())
+
+    def has_error(self):
+        return self.error and (not isinstance(self.error, RqlCursorEmpty))
+
+    def is_empty(self):
+        return isinstance(self.error, RqlCursorEmpty) and len(self.items) == 0
+
+    def _get_next(self, timeout):
+        if self.is_empty() or self.has_error():
```

Comment:
This shouldn't fail until all successful rows have been read out from the cursor - that is, when `len(self.items) == 0`.  Only once all the items have been read should we throw the error.